# Partisan Voter Index (PVI) calculations for Massachusetts Legislative Districts

The Partisan Voter Index or PVI was created by the Cook Political Report to create a common metric for comparing the partisanship or Democratic/Republican lean of legislative districts. The Cook Political Report primarily uses the PVI for comparing U.S. House districts, but it can also be used to compare state legislative districts.

PVI is calculated by taking the Democratic percentage of the over the last two presidential elections for the area you are considering and then subtracting the average for the U.S. as a whole.

In [0]:
from __future__ import division

US_VOTES = {
    "Obama_12": 65915796,
    "Romney_12": 60933500,
    "Clinton_16": 65853516,
    "Trump_16": 62984824,
    }

def dem_pct(d, r):
    """The Democratic percentage of the vote, given Dem and GOP vote totals."""
    return (d / (d + r))

US_DEM_PCT = dem_pct(US_VOTES["Obama_12"] + US_VOTES["Clinton_16"],
                     US_VOTES["Romney_12"] + US_VOTES["Trump_16"])

def calc_pvi(df):
    """Numeric PVI given vote totals from last two presidential elections."""
    dem = df["Obama_12"] + df["Clinton_16"]
    gop = df["Romney_12"] + df["Trump_16"]
    pvi = (dem_pct(dem, gop) - US_DEM_PCT) * 100
    return pvi

def pvi_string(pvi):
    """String representative of numeric PVI value."""
    if pvi <= -0.5:
        s = "R+{:.0f}".format(abs(pvi))
    elif pvi >= 0.5:
        s = "D+{:.0f}".format(abs(pvi))
    else:
        s = "EVEN"
    return s

For example, we can calculate the PVI for Massachusetts as a whole using the vote totals from the 2012 and 2016 presidential elections.

In [2]:
MA_VOTES = {
    "Obama_12": 1921290,
    "Romney_12": 1188314,
    "Clinton_16": 1995196,
    "Trump_16": 1090893,
    }

MA_PVI_N = calc_pvi(MA_VOTES)
MA_PVI_N

11.677772318663537

So Massachusetts is about 11.7 points more Democratic than the country as a whole using the PVI metric.

There is a formatting convention for PVI to round the numeric value to an integer and show negative/GOP-leaning PVI with an "R+" and positive/Dem-leaning PVI with a "D+".

In [3]:
MA_PVI = pvi_string(MA_PVI_N)
MA_PVI

'D+12'

## Precinct-level presidential results

In order to calculate PVI across various legislative districts in Massachusetts, we will start by downloading the precinct-level presidential election data for 2012 and 2016 from the awesome [electionstats.state.ma.us](http://electionstats.state.ma.us) website created by Adam Friedman and supported by the Massachusetts Secretary of the Commonwealth.

Once we have the precinct-level data, we can map the precincts to legislative districts and then do the PVI calculations.

In [4]:
import pandas as pd

def read_pd43(url, set_index=True):
    """Read a precinct-level CSV file from PD43 site."""
    # Force Ward and Pct to string for consistency.
    p = pd.read_csv(url, dtype={"Ward": str, "Pct": str}, thousands=",")
    # Remove TOTALS row
    p = p[p["City/Town"] != "TOTALS"]
    # Optionally set the index to three precinct-identifying columns
    if set_index:
        p = p.set_index(["City/Town", "Ward", "Pct"])
    return p

p12 = read_pd43("http://electionstats.state.ma.us/elections/download/22515/precincts_include:1/")
p12.head(20)

Obama/ Biden  Romney/ Ryan  Johnson/ Gray  Stein/ Honkala  \
City/Town Ward Pct                                                              
Abington  -    1             844           705              6               7   
               2             742           741             18               8   
               3             815           801             13              11   
               4             899           904              8               9   
               5             844           910              8               2   
Acton     -    1            1228           583             18               5   
               2            1216           722             19              14   
               3            1378           699             34              24   
               4            1462           636             25              20   
               5            1454           565             25              11   
               6            1134           670             15               9   
Acushnet  -    1            1034           574             17               8   
               2             959           803             25              11   
               3            1053           819             17               7   
Adams     -    1             525           164             17               1   
               2             607           172              9               6   
               3             582           169             10               3   
               4             590           129              9               6   
               5             709           185              8               9   
Agawam    -    1             850           665             13               7   

                    All Others  No Preference  Blank Votes  Total Votes Cast  
City/Town Ward Pct                                                            
Abington  -    1             4              0            5              1571  
               2             4              0            7              1520  
               3             2              0            6              1648  
               4             5              0            5              1830  
               5             5              0            5              1774  
Acton     -    1             3              0            8              1845  
               2             3              0            4              1978  
               3             0              0            8              2143  
               4             0              0            6              2149  
               5             8              0            8              2071  
               6             3              0            5              1836  
Acushnet  -    1            13              0           23              1669  
               2             4              0           19              1821  
               3             1              0           11              1908  
Adams     -    1             0              0            1               708  
               2             2              0            5               801  
               3             2              0            2               768  
               4             3              0            3               740  
               5             2              0            1               914  
Agawam    -    1             6              0            5              1546

We are only interested in the Democratic and Republican vote totals.

In [5]:
p12 = p12.rename(columns={"Obama/ Biden": "Obama_12", "Romney/ Ryan": "Romney_12"})[["Obama_12", "Romney_12"]]
p12.head(20)

Obama_12  Romney_12
City/Town Ward Pct                     
Abington  -    1         844        705
               2         742        741
               3         815        801
               4         899        904
               5         844        910
Acton     -    1        1228        583
               2        1216        722
               3        1378        699
               4        1462        636
               5        1454        565
               6        1134        670
Acushnet  -    1        1034        574
               2         959        803
               3        1053        819
Adams     -    1         525        164
               2         607        172
               3         582        169
               4         590        129
               5         709        185
Agawam    -    1         850        665

Repeat the same process for 2016 presidential result data.

In [6]:
p16 = read_pd43("http://electionstats.state.ma.us/elections/download/40060/precincts_include:1/")
p16 = p16.rename(columns={"Clinton/ Kaine": "Clinton_16", "Trump/ Pence": "Trump_16"})[["Clinton_16", "Trump_16"]]
p16.head(20)

Clinton_16  Trump_16
City/Town Ward Pct                      
Abington  -    1           818       717
               2           739       785
               3           773       808
               4           877       878
               5           908       829
Acton     -    1          1459       438
               2          1550       480
               3          1580       435
               4          1675       414
               5          1580       387
               6          1354       401
Acushnet  -    1           783       786
               2           756      1043
               3           862      1005
Adams     -    1           439       255
               2           476       270
               3           475       268
               4           446       233
               5           577       259
Agawam    -    1           631       827

Now we combine the 2012 and 2016 data together.

In [7]:
p = p12.join(p16)
p.head(20)

Obama_12  Romney_12  Clinton_16  Trump_16
City/Town Ward Pct                                           
Abington  -    1         844        705         818       717
               2         742        741         739       785
               3         815        801         773       808
               4         899        904         877       878
               5         844        910         908       829
Acton     -    1        1228        583        1459       438
               2        1216        722        1550       480
               3        1378        699        1580       435
               4        1462        636        1675       414
               5        1454        565        1580       387
               6        1134        670        1354       401
Acushnet  -    1        1034        574         783       786
               2         959        803         756      1043
               3        1053        819         862      1005
Adams     -    1         525        164         439       255
               2         607        172         476       270
               3         582        169         475       268
               4         590        129         446       233
               5         709        185         577       259
Agawam    -    1         850        665         631       827

## Mapping precincts to legislative districts

In order to map the precincts to legislative districts we download results from all of the elections for that office for the year 2016 and mark each one with the district name. This requires some web page scraping using the BeautifulSoup HTML parsing library. 

In [8]:
import bs4
import requests

ELECTION_QUERY = "http://electionstats.state.ma.us/elections/search/year_from:{year}/year_to:{year}/office_id:{office_id}/stage:{stage}"
STATE_REP_ID = 8
STATE_SEN_ID = 9
US_CONG_ID = 5

def get_election_ids(year, office_id, stage):
    """Returns a list (ELECTION_ID, DISTRICT) for the given query parameters."""
    r = requests.get(ELECTION_QUERY.format(year=year, office_id=office_id, stage=stage))
    bs = bs4.BeautifulSoup(r.text, "html5lib")
    elec_trs = bs.findAll("tr", {"class": "election_item general_party"})
    elec_ids = [district_and_id(elec_tr) for elec_tr in elec_trs]
    return elec_ids
  
def district_and_id(elec_tr):
    """Parse the ELECTION_ID and DISTRICT from the HTML element."""
    election_id = elec_tr["id"].split("-")[2]
    district = elec_tr.td.findNext("td").findNext("td").text
    return (election_id, district)

CSV_QUERY = "http://electionstats.state.ma.us/elections/download/{election_id}/precincts_include:1/"

def get_office_precincts(year, office_id, stage):
    """Query for the matching elections, return all of the precincts tagged with the district name."""
    election_list = []
    election_ids = get_election_ids(year, office_id, stage)
    for election_id, district in election_ids:
        p = read_pd43(CSV_QUERY.format(election_id=election_id), set_index=False)
        p = p[["City/Town", "Ward", "Pct"]]
        p["District"] = district
        election_list.append(p)
    all_precincts = pd.concat(election_list, ignore_index=True).drop_duplicates()
    all_precincts = all_precincts.set_index(["City/Town", "Ward", "Pct"])
    return all_precincts

sr_pcts = get_office_precincts(2016, STATE_REP_ID, "General").rename(columns={"District": "State_Rep"})
ss_pcts = get_office_precincts(2016, STATE_SEN_ID, "General").rename(columns={"District": "State_Sen"})
ush_pcts = get_office_precincts(2016, US_CONG_ID, "General").rename(columns={"District": "US_House"})

# Combine the presidential result data with all of the district tags
pvi = p.join(ush_pcts).join(ss_pcts).join(sr_pcts).reset_index()
pvi.head(20)

City/Town Ward Pct  Obama_12  Romney_12  Clinton_16  Trump_16  \
0   Abington    -   1       844        705         818       717   
1   Abington    -   2       742        741         739       785   
2   Abington    -   3       815        801         773       808   
3   Abington    -   4       899        904         877       878   
4   Abington    -   5       844        910         908       829   
5      Acton    -   1      1228        583        1459       438   
6      Acton    -   2      1216        722        1550       480   
7      Acton    -   3      1378        699        1580       435   
8      Acton    -   4      1462        636        1675       414   
9      Acton    -   5      1454        565        1580       387   
10     Acton    -   6      1134        670        1354       401   
11  Acushnet    -   1      1034        574         783       786   
12  Acushnet    -   2       959        803         756      1043   
13  Acushnet    -   3      1053        819         862      1005   
14     Adams    -   1       525        164         439       255   
15     Adams    -   2       607        172         476       270   
16     Adams    -   3       582        169         475       268   
17     Adams    -   4       590        129         446       233   
18     Adams    -   5       709        185         577       259   
19    Agawam    -   1       850        665         631       827   

             US_House                          State_Sen       State_Rep  
0   8th Congressional               Norfolk and Plymouth    7th Plymouth  
1   8th Congressional               Norfolk and Plymouth    7th Plymouth  
2   8th Congressional               Norfolk and Plymouth    7th Plymouth  
3   8th Congressional               Norfolk and Plymouth    7th Plymouth  
4   8th Congressional               Norfolk and Plymouth    7th Plymouth  
5   3rd Congressional            Middlesex and Worcester  14th Middlesex  
6   3rd Congressional            Middlesex and Worcester  14th Middlesex  
7   3rd Congressional            Middlesex and Worcester  37th Middlesex  
8   3rd Congressional            Middlesex and Worcester  37th Middlesex  
9   3rd Congressional            Middlesex and Worcester  37th Middlesex  
10  3rd Congressional            Middlesex and Worcester  14th Middlesex  
11  9th Congressional           2nd Bristol and Plymouth    11th Bristol  
12  9th Congressional           2nd Bristol and Plymouth    11th Bristol  
13  9th Congressional           2nd Bristol and Plymouth    11th Bristol  
14  1st Congressional  Berkshire, Hampshire and Franklin   1st Berkshire  
15  1st Congressional  Berkshire, Hampshire and Franklin   1st Berkshire  
16  1st Congressional  Berkshire, Hampshire and Franklin   1st Berkshire  
17  1st Congressional  Berkshire, Hampshire and Franklin   1st Berkshire  
18  1st Congressional  Berkshire, Hampshire and Franklin   1st Berkshire  
19  1st Congressional          2nd Hampden and Hampshire     3rd Hampden

## Calculate PVI for each office/district

For each office—State Representative, State Senate, and U.S. House—group by district, summing up the presidential votes and then calculate the numeric PVI and PVI string.

Begin with State Representative.

In [9]:
sr_pvi = pvi.groupby("State_Rep").sum().reset_index()
sr_pvi["PVI_N"] = calc_pvi(sr_pvi)
sr_pvi["PVI"] = sr_pvi["PVI_N"].map(pvi_string)
sr_pvi.sort_values("PVI_N", ascending=False)

State_Rep  Obama_12  Romney_12  Clinton_16  Trump_16      PVI_N  \
133     6th Suffolk     15073        644       15327       744  44.098303   
125     5th Suffolk     13404        681       13467       736  43.455539   
14     11th Suffolk     17136       1036       18292       920  43.232546   
141     7th Suffolk     10657        806       10230       574  42.267219   
72   25th Middlesex     16802       1778       17350       846  41.329643   
22     12th Suffolk     17172       1923       16953      1673  38.931579   
2      10th Hampden     10931       1200       10464      1266  38.129875   
39     15th Suffolk     15597       2258       16631      1556  37.882632   
74   27th Middlesex     17118       2580       19929      1990  37.483641   
73   26th Middlesex     14454       2212       16361      1911  36.663826   
157     9th Suffolk     12661       2410       14003      1336  36.146414   
10     11th Hampden     12442       1738       11331      1843  35.373409   
41       16th Essex     10072       1579       11076      1679  35.115554   
38     15th Norfolk     15638       3597       17136      1885  34.134952   
102   3rd Hampshire     14569       2820       14054      2155  33.657302   
47     17th Suffolk     10924       2658       13150      1848  32.698460   
76   29th Middlesex     17345       4433       18931      3532  30.461069   
93   34th Middlesex     14397       3656       16103      3343  29.800233   
156    9th Plymouth     10311       2317       10544      2512  29.663142   
149     8th Suffolk     14271       4882       16708      2349  29.540365   
34     14th Suffolk     14593       3692       15602      3371  29.507728   
51     18th Suffolk      9510       2955       10740      1905  29.109892   
11   11th Middlesex     15352       5482       17570      3553  26.930779   
106     3rd Suffolk     12835       4938       16034      3109  26.666594   
19   12th Middlesex     15065       5748       17005      3404  26.262994   
61    1st Hampshire     17346       5069       17368      4919  26.121211   
28     13th Suffolk     13271       4078       13805      3778  25.975323   
98    3rd Berkshire     14513       3750       12927      4405  25.554209   
71   24th Middlesex     16938       6519       18931      4350  25.209564   
87      2nd Suffolk     10443       3911       12466      3144  24.919810   
..              ...       ...        ...         ...       ...        ...   
80        2nd Essex     12584      13456       13376     11554  -0.603349   
134   6th Worcester      9644       8321        8485      9180  -0.653990   
30     14th Bristol      9526      10027       10206      9064  -0.709728   
36       15th Essex     10185      10039       10613     10089  -0.716719   
115    4th Plymouth     11677      12654       12723     10983  -0.741090   
62    1st Middlesex     10921      11742       11685     10326  -0.933129   
82      2nd Hampden     11352      11569       11453     10711  -0.953036   
112     4th Hampden      9281       8218        8225      9021  -1.151041   
25       13th Essex     11523      11668       11727     11288  -1.217129   
142   7th Worcester     10188       9872        9691     10153  -1.718208   
86     2nd Plymouth     10669       9808        9510     10558  -1.765877   
155     9th Norfolk     10458      12604       12036     10193  -1.869773   
66    1st Worcester     11427      12662       12098     11222  -1.913889   
69   22nd Middlesex      9839       9959       10100     10319  -1.956733   
110     4th Bristol     10336      10055        9911     10973  -2.481363   
67   20th Middlesex     11417      13917       12885     11877  -3.024410   
53   19th Middlesex     10665      11387       10682     11299  -3.055720   
140    7th Plymouth     10323      10669        9921     10903  -3.123178   
150   8th Worcester     10523      10702        9773     10958  -3.160782   
60      1st Hampden     10356       9658        8783     10816  -3.220322   
21    

Next, State Senate.

In [10]:
ss_pvi = pvi.groupby("State_Sen").sum().reset_index()
ss_pvi["PVI_N"] = calc_pvi(ss_pvi)
ss_pvi["PVI"] = ss_pvi["PVI_N"].map(pvi_string)
ss_pvi.sort_values("PVI_N", ascending=False)

State_Sen  Obama_12  Romney_12  \
17                                  2nd Suffolk     61429       5511   
7                                   1st Suffolk     55653      12790   
29                        Middlesex and Suffolk     48613      12190   
14                                2nd Middlesex     65521      18456   
5                     1st Middlesex and Norfolk     59893      22633   
18                    2nd Suffolk and Middlesex     48768      18089   
28            Hampshire, Franklin and Worcester     58698      17217   
27                                      Hampden     42046      12097   
24            Berkshire, Hampshire and Franklin     58586      18448   
8                     1st Suffolk and Middlesex     44155      18341   
32                          Norfolk and Suffolk     52226      31550   
21                                3rd Middlesex     53154      34929   
10                     2nd Bristol and Plymouth     45231      20642   
15                    2nd Middlesex and Norfolk     45965      30724   
9                                 1st Worcester     43226      25973   
22                                4th Middlesex     54733      35193   
30                      Middlesex and Worcester     51137      36090   
23                                5th Middlesex     49814      33149   
20                                    3rd Essex     46191      29139   
34                Norfolk, Bristol and Plymouth     49390      34573   
12                      2nd Essex and Middlesex     42198      28960   
4                                 1st Middlesex     39003      27591   
11                                    2nd Essex     49958      34836   
16                     2nd Plymouth and Bristol     41169      27852   
26                             Cape and Islands     60846      45455   
31                         Norfolk and Plymouth     43375      31328   
0                      1st Bristol and Plymouth     43729      25013   
13                    2nd Hampden and Hampshire     43416      28425   
33               Norfolk, Bristol and Middlesex     45902      39855   
19                                2nd Worcester     42739      33091   
1                                     1st Essex     46384      36992   
3                     1st Hampden and Hampshire     44977      34037   
2                       1st Essex and Middlesex     49983      48049   
37                      Worcester and Middlesex     37951      33727   
36                         Plymouth and Norfolk     47407      47214   
25                          Bristol and Norfolk     41730      39923   
35                      Plymouth and Barnstable     44779      43122   
6                      1st Plymouth and Bristol     42005      37485   
38                        Worcester and Norfolk     40023      38344   
39  Worcester, Hampden, Hampshire and Middlesex     39287      39281   

    Clinton_16  Trump_16      PVI_N   PVI  
17       64278      4208  41.288118  D+41  
7        59805     11511  31.076942  D+31  
29       53276     10312  30.374997  D+30  
14       74590     15118  29.134334  D+29  
5        68176     13300  26.554632  D+27  
18       55191     12163  25.924104  D+26  
28       55922     17212  25.365616  D+25  
27       38681     13463  24.416634  D+24  
24       52576     21547  22.005486  D+22  
8        49333     17278  20.875986  D+21  
32       58690     24888  14.741007  D+15  
21       60180     24943  13.897799  D+14  
10       38188     25543  12.829256  D+13  
15       52060     23699  12.765345  D+13  
9        45184     23463  12.601521  D+13  
22       58689     30974  11.621160  D+12  
30       57880     28082  11.411569  D+11  
23       54639     28649  11.293225  D+11  
20       46799     27654  10.547878  D+11  
34       52644     28984  10.082814  D+10  
12       45321     26808   9.544242  D+10  
4        42273     24530   9.392638   D+9  
11       52382     32448   8.798174   D+9  
16       42027     27772   8.

And finally, U.S. House.

In [11]:
ush_pvi = pvi.groupby("US_House").sum().reset_index()
ush_pvi["PVI_N"] = calc_pvi(ush_pvi)
ush_pvi["PVI"] = ush_pvi["PVI_N"].map(pvi_string)
ush_pvi.sort_values("PVI_N", ascending=False)

US_House  Obama_12  Romney_12  Clinton_16  Trump_16      PVI_N  \
6  7th Congressional    233382      44275      254037     36018  34.321468   
4  5th Congressional    235984     119934      258908     95922  18.094473   
0  1st Congressional    213423     114339      194036    123953  11.563200   
7  8th Congressional    213364     150825      231356    131624   9.622453   
3  4th Congressional    211423     152699      225976    133705   8.895400   
1  2nd Congressional    199549     133195      197492    129437   8.652278   
2  3rd Congressional    189461     137869      202952    123347   8.500776   
5  6th Congressional    212003     169966      224858    153244   5.941072   
8  9th Congressional    212701     165212      205581    163643   4.449376   

    PVI  
6  D+34  
4  D+18  
0  D+12  
7  D+10  
3   D+9  
1   D+9  
2   D+9  
5   D+6  
8   D+4